bberg/src/lookup_builder.rs

```diff
@@ -0,0 +1,366 @@
+use crate::{
+    file_writer::BBFiles,
+    utils::{create_get_const_entities, create_get_nonconst_entities},
+};
+use ast::{
+    analyzed::{AlgebraicExpression, Analyzed, Identity, IdentityKind},
+    parsed::SelectedExpressions,
+};
+use itertools::Itertools;
+use number::FieldElement;
+
+use crate::utils::sanitize_name;
+
+#[derive(Debug)]
+/// Lookup
+///
+/// Contains the information required to produce a lookup relation
+pub struct Lookup {
+    ///  the name given to the inverse helper column
+    pub attribute: Option<String>,
+    /// The name of the counts polynomial that stores the number of times a lookup is read
+    pub counts_poly: String,
+    /// the left side of the lookup
+    pub left: LookupSide,
+    /// the right side of the lookup
+    pub right: LookupSide,
+}
+
+#[derive(Debug)]
+/// PermSide
+///
+/// One side of a two sided lookup relationship
+pub struct LookupSide {
+    /// -> Option<String> - the selector for the lookup ( on / off toggle )
+    selector: Option<String>,
+    /// The columns involved in this side of the lookup
+    cols: Vec<String>,
+}
+
+pub trait LookupBuilder {
+    /// Takes in an AST and works out what lookup relations are needed
+    /// Note: returns the name of the inverse columns, such that they can be added to he prover in subsequent steps
+    fn create_lookup_files<F: FieldElement>(
+        &self,
+        name: &str,
+        analyzed: &Analyzed<F>,
+    ) -> Vec<Lookup>;
+}
+
+impl LookupBuilder for BBFiles {
```

Comment:
In this function and subsequent ones, replace variables like "perm", "perm_settings" by "lookup", "lookup_settings".